feat: add support for accessor arrays and refactor `stats/base/cumax`

vivekmaurya001 opened the pull request:
Resolves #{{TODO: add issue number}}.

This pull request:

-    add accessor arrays support and refactor `stats/base/cumax`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #{{TODO: add issue number}} 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

 - [x] Read, understood, and followed the [contributing guidelines](https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md)


* * *

@stdlib-js/reviewers 

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

vivekmaurya001 committed:
add support for accessor arrays

vivekmaurya001 committed:
Merge branch 'stat/cumax' of https://github.com/vivekmaurya001/stdlib into stat/cumax

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

vivekmaurya001 committed:
add support for accessor arrays

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

vivekmaurya001 committed:
add support for accessor arrays

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

vivekmaurya001 committed:
add support for accessor arrays

vivekmaurya001 committed:
add support for accessor arrays

vivekmaurya001 committed:
add support for accessor arrays

vivekmaurya001 commented:
/stdlib lint-autofix

stdlib-bot committed:
fix: resolve lint errors

vivekmaurya001 committed:
add support for accessor arrays

vivekmaurya001 committed:
Merge branch 'stat/cumax' of https://github.com/vivekmaurya001/stdlib into stat/cumax

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

vivekmaurya001 committed:
add support for accessor arrays

vivekmaurya001 committed:
add support for accessor arrays

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

vivekmaurya001 committed:
add support for accessor arrays

vivekmaurya001 commented:
@kgryte , I have tried to remove the error occuring in accessor file example  , can u please take a look how to resolve error

kgryte commented on lib/node_modules/@stdlib/stats/base/cumax/lib/accessors.js:
```diff
@@ -0,0 +1,111 @@
+* @example
+* var toAccessorArray = require( '@stdlib/array/base/to-accessor-array' );
+*
+* var x = toAccessorArray( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ] );
+* var y = toAccessorArray( [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ] );
+*
+* cumax( 4, x, 2, 1, y, 1, 0 );
+*
+* console.log( y );
+* // => toAccessorArray( [ 1.0, 1.0, 2.0, 4.0, 0.0, 0.0, 0.0, 0.0 ] );
```
```suggestion
* var x = [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ];
* var y = [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ];
*
* cumax( 4, toAccesorArray( x ), 2, 1, toAccessorArray( y ), 1, 0 );
* // y => [ 1.0, 1.0, 2.0, 4.0, 0.0, 0.0, 0.0, 0.0 ];
```

vivekmaurya001 replied:
i have changed it but still getting the error

vivekmaurya001 committed:
add support for accessor arrays

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

vivekmaurya001 committed:
add support for accessor arrays

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on lib/node_modules/@stdlib/stats/base/cumax/lib/accessors.js:
```diff
@@ -0,0 +1,109 @@
+* @returns {NumericArray} output array
+*
+* @example
+* var toAccessorArray = require( '@stdlib/array/base/to-accessor-array' );
+*
+* var x = [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ];
+* var y = [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ];
+*
+* cumax( 4, toAccessorArray( x ), 2, 1, toAccessorArray( y ), 1, 0 );
+* // y => [ 1.0, 1.0, 2.0, 4.0, 0.0, 0.0, 0.0, 0.0 ];
```
```suggestion
* var arraylike2object = require( '@stdlib/array/base/arraylike2object' );
* var toAccessorArray = require( '@stdlib/array/base/to-accessor-array' );
*
* var x = [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ];
* var y = [ 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 ];
*
* cumax( 4, arraylike2object( toAccessorArray( x ) ), 2, 1, arraylike2object( toAccessorArray( y ) ), 1, 0 );
* // y => [ 1.0, 1.0, 2.0, 4.0, 0.0, 0.0, 0.0, 0.0 ];
```

For future reference, https://github.com/stdlib-js/stdlib/blob/ef0c134d557fa7364dd80c9eaa40d2df9189f4b1/lib/node_modules/%40stdlib/blas/ext/base/gcusumors/lib/accessors.js

kgryte replied:
Also, just look at the implementation (see `./ndarray.js` and what is being passed in).

vivekmaurya001 replied:
ook

kgryte commented on lib/node_modules/@stdlib/stats/base/cumax/lib/accessors.js:
```diff
@@ -0,0 +1,109 @@
+*
+* @private
+* @param {PositiveInteger} N - number of indexed elements
+* @param {NumericArray} x - input array
```
This isn't correct.

vivekmaurya001 replied:
Yess , i will correct it using reference soon

vivekmaurya001 committed:
add support for accessor arrays

vivekmaurya001 commented:
@kgryte , thanks for help , now it is ready for review

vivekmaurya001 committed:
add support for accessor arrays

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

vivekmaurya001 commented:
@kgryte , i have reviewed it again and maded some undone changes from reference

kgryte commented on lib/node_modules/@stdlib/stats/base/cumax/README.md:
```diff
@@ -52,11 +52,11 @@ The function has the following parameters:
 -   **N**: number of indexed elements.
 -   **x**: input [`Array`][mdn-array] or [`typed array`][mdn-typed-array].
--   **strideX**: index increment for `x`.
+-   **strideX**: Stride length for `x`.
```
```suggestion
-   **strideX**: stride length for `x`.
```

kgryte commented on lib/node_modules/@stdlib/stats/base/cumax/README.md:
```diff
@@ -52,11 +52,11 @@ The function has the following parameters:
+-   **strideX**: Stride length for `x`.
 -   **y**: output [`Array`][mdn-array] or [`typed array`][mdn-typed-array].
--   **strideY**: index increment for `y`.
+-   **strideY**: Stride length for `y`.
```
```suggestion
-   **strideY**: stride length for `y`.
```

kgryte commented on lib/node_modules/@stdlib/stats/base/cumax/examples/index.js:
```diff
@@ -18,22 +18,16 @@
 console.log( y );
 
-cumax( x.length, x, 1, y, -1 );
+cumax( x.length, x, 1, y, 1 );
```
```suggestion
cumax( x.length, x, 1, y, -1 );
```
Ensure that README and this example match.

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/stats/base/cumax/README.md:
```diff
@@ -122,6 +122,7 @@ cumax.ndarray( 4, x, 2, 1, y, -1, y.length-1 );
 
 -   If `N <= 0`, both functions return `y` unchanged.
 -   Depending on the environment, the typed versions ([`dcumax`][@stdlib/stats/strided/dcumax], [`scumax`][@stdlib/stats/base/scumax], etc.) are likely to be significantly more performant.
+-   Both functions support array-like objects having getter and setter accessors for array element access (e.g., [`@stdlib/array/base/accessor`][@stdlib/array/base/accessor])
```
```suggestion
-   Both functions support array-like objects having getter and setter accessors for array element access (e.g., [`@stdlib/array/base/accessor`][@stdlib/array/base/accessor]).
```

kgryte committed:
docs: fix missing period

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM. Thanks, @vivekmaurya001!
